### PR TITLE
`asArray` macro implementation

### LINKED ISSUE #6563: Implement an `asArray` macro
Inspired by this: https://github.com/nim-lang/Nim/pull/6301/files#diff-15d82a47d4b87ec3cce5f6262d3828ceR339

```nim
var x = [
  12.Rune,
  13.Rune,
  14.Rune,
  ...
]
```

Makes me want to cry.

So let's allow this:

```nim
var x = asArray[Rune]([
  12,
  13,
  14,
  ...
])
```

(Or similar)

And add it to the stdlib! (possibly to ``sequtils``)

### PR DESCRIPTION
**Pull request by couven92**
fixes #6563

### CODE REVIEW

**Comment by couven92** [comment]
according to @Araq creating generic macros and using the generic type parameter using `getType(T)` is not currently possible in Nim.

Therefore, the `asArray` macro takes the `typedesc` of the type to cast to as last parameter.

The target type parameter is last, as that makes it consistent with the `array[L..H, Type]` type signature (where the array element type occurs last)

File: lib/pure/collections/sequtils.nim [hunk 1]
```diff
@@ -704,6 +704,31 @@ template newSeqWith*(len: int, init: untyped): untyped =
     result[i] = init
   result
 
+when not defined(nimscript):
+  import macros
+  macro asArray*(values: typed, targetType: typedesc): untyped =
+    ## converts a static array to an array of type ``targetType`` by converting
```
Comment (dom96) [inline lib/pure/collections/sequtils.nim]: I think you should change this to something like this:

> applies a type conversion to each of the elements in the specified array literal. Each element is converted to the ``targetType`` type.

**Comment by dom96** [comment]
Thank you for implementing this :D

File: lib/pure/collections/sequtils.nim [hunk 2]
```diff
@@ -704,6 +704,31 @@ template newSeqWith*(len: int, init: untyped): untyped =
     result[i] = init
   result
 
+when not defined(nimscript):
```
Comment (dom96) [inline lib/pure/collections/sequtils.nim]: Why is it a problem for nimscript?

File: lib/pure/collections/sequtils.nim (hunk 1 above)
Comment (couven92) [reply lib/pure/collections/sequtils.nim]: @dom96 Error from nim while executing `nim e tests\test_nimscript.nims`
> lib\core\macros.nim(735, 1) Error: cannot generate VM code for macro dumpTree(s: untyped): untyped

File: lib/pure/collections/sequtils.nim (hunk 2 above)
Comment (couven92) [reply lib/pure/collections/sequtils.nim]: @dom96 Error from nim while executing `nim e tests\test_nimscript.nims`
> lib\core\macros.nim(735, 1) Error: cannot generate VM code for macro dumpTree(s: untyped): untyped

Comment (dom96) [reply lib/pure/collections/sequtils.nim]: seems like a bug(?)

**Comment by data-man** [comment]
Can it be better to swap parameters?
```nim
macro asArray*(targetType: typedesc, values: typed): untyped
```

**Comment by data-man** [comment]
And move this to macros.nim?

**Comment by couven92** [comment]
@data-man Hmm... The parameter ordering is explained here.

As for putting it in `macros.nim`: I don't know... I think it fits best in `sequtils` as @dom96 suggested, but I don't really care

@dom96 what do you think?

**Comment by data-man** [comment]
@couven92: Because to declare an array you must import sequtils.

**Comment by couven92** [comment]
@data-man, no only if you want Nim to do magic casting for you. And sequtils fits best.

actually disregard me saying I don't care, we're not putting this to macros, that would be totally out of place there

**Comment by data-man** [comment]
Besides, it would be the first and only macro in sequtils. :)

**Comment by couven92** [comment]
Nothing wrong with that, is there?

**Comment by data-man** [comment]
I just expressed my opinion.
In principle, no difference what module to import: macros or sequtils.

**Comment by couven92** [comment]
Yeah, I hear you. But the macros module is really about doing operations on NimNodes, asArray does not do anything with AST nodes, it deals with static array declarations. And the module that is closest in the standard library that deals with quality-of-life-improvements for sequences and arrays is sequtils.

**Comment by data-man** [comment]
@couven92 
In any case, I now use your macro and very happy.
Thank you!

**Comment by data-man** [comment]
About the swap of parameters.
When a variable is declared, its type is first given and then the value.
It seems more consistent to make your macro look like this.

**Comment by couven92** [comment]
I agree

**Comment by Araq** [comment]
Lets put it into the future module.

**Comment by data-man** [comment]
@dom96 @Araq 

Ready to merge?
I'm preparing a PR, which depends on this.
Although it would be ideal if the compiler would allow this code:
```nim
var
  a: array[3, uint8] = [1, 2, 3]
```
Now it is allowed to write only so:
```nim
  a: array[3, uint8] = [1'u8, 2'u8 , 3'u8]
```
Or so:
```nim
  a: array[3, uint8] = [1'u8, 2, 3]
```

**Comment by couven92** [comment]
I say **ready to merge**

File: lib/pure/future.nim [hunk 3]
```diff
@@ -197,4 +197,32 @@ macro dump*(x: typed): untyped =
...
+  ##
+  ## .. code-block::
+  ##   let x = asArray(int, [0.1, 1.2, 2.3, 3.4])
+  ##   doAssert x is array[4, int]
+  ##
+  ## Short notation for:
+  ##
+  ## .. code-block::
+  ##   let x = [(0.1).int, (1.2).int, (2.3).int, (3.4).int]
+  let tNode = getType(targetType)[1]
```
Comment (Araq) [inline lib/pure/future.nim]: Don't use ``getType`` here. Use a purely syntactic macro.

File: lib/pure/future.nim [hunk 4]
```diff
@@ -197,4 +197,32 @@ macro dump*(x: typed): untyped =
...
+  values.expectKind(nnkBracket)
+  result = newNimNode(nnkBracket, lineInfoFrom=values)
+  for i in 0 ..< len(values):
+    var dot = newNimNode(nnkDotExpr, lineInfoFrom=values[i])
```
Comment (Araq) [inline lib/pure/future.nim]: Do not produce a dot expression for the conversion but a call expression.

Comment (couven92) [reply lib/pure/future.nim]: done

File: lib/pure/future.nim (hunk 3 above)
Comment (couven92) [reply lib/pure/future.nim]: @Araq sorry, I have no idea what you mean. Don't I need `getType`... else I got `targetType` in the produced code. I am **very probably** doing something fundamentally wrong, but I don't understand how else I should do this :(

Comment (couven92) [reply lib/pure/future.nim]: maybe you mean like I did here: https://github.com/nim-lang/Nim/pull/6640/commits/31ace8ad297731cf2f9f43f995a3cc794b3d77fc ?

**Comment by dom96** [comment]
Into the ``future`` module? Does this imply it will end up in system?

**Comment by Araq** [comment]
> Into the future module? Does this imply it will end up in system?

No, I'd like to rename ``future`` to ``sugar.nim`` or similar.

**Comment by Araq** [comment]
> Into the future module? Does this imply it will end up in system?

Well the ``=>`` and ``->`` can be moved to system, but ``lc`` is not yet production ready and so IMO we should keep it in a new module called ``sugar``.

**Comment by couven92** [comment]
Okay @Araq, I moved `asArray` and `ListComprehension` into a new `lib/pure/sugar.nim`. (https://github.com/nim-lang/Nim/pull/6640/commits/9bdee897747d3345d1c7722b3440d7cf38460f40)

**Comment by dom96** [comment]
Sorry, but IMO asArray should be in ``sequtils`` if it won't be in ``system``. I would honestly prefer it in ``sequtils``.

**Comment by Araq** [comment]
Rebase to make the tests green.

File: lib/pure/collections/sequtils.nim [hunk 5]
```diff
@@ -706,7 +706,7 @@ template newSeqWith*(len: int, init: untyped): untyped =
     result[i] = init
   result
 
-macro asArray*(targetType: untyped, values: typed): untyped =
+macro asArray*(targetType: typedesc, values: typed): untyped =
```
Comment (couven92) [inline lib/pure/collections/sequtils.nim]: @Araq, yes that was obviously my first attempt as well. However, this does not work. When I attempt to add `targetType` to the result NimNode, i get the following error from the compiler:
> sequtils.nim(727, 9) Error: type mismatch: got (NimNode, typedesc)

The only way, I can see, to fix this is to use `getType` on `targetType` to get the symbol that is bound to the target type. However, you did not want me to do that.

Comment (Araq) [reply lib/pure/collections/sequtils.nim]: I have a solution for this that I want to try out.

**Comment by Araq** [comment]
Replaced by ``mapLiterals``.

**Pull request closed without merging**